File: scripts/PreProcessMatrix.py
import argparse
import numpy as np
import scipy.stats as st
import os
import sys
import matplotlib.pyplot as plt
import timeit
# import stuff in the src folder
try:
	import MRS
except ImportError:
	sys.path.append(os.path.join(os.path.dirname(os.path.dirname(os.path.abspath(__file__))), 'src'))
	import MRS
try:
	import SS
except ImportError:
	sys.path.append(os.path.join(os.path.dirname(os.path.dirname(os.path.abspath(__file__))), 'src'))
	import SS
try:
	import NumSwitch
except ImportError:
	sys.path.append(os.path.join(os.path.dirname(os.path.dirname(os.path.abspath(__file__))), 'src'))
	import NumSwitch

if __name__ == '__main__':
	parser = argparse.ArgumentParser(description="This script all indices from the Koslicki & Novak (2018) paper", formatter_class=argparse.ArgumentDefaultsHelpFormatter)
	parser.add_argument('input_file', type=str, help="Input comma separated file for the jacobian matrix.")
	parser.add_argument('output_folder', type=str, help="Output folder. A number of files will be created in the form 'output_folder/<prefix>_*.npy'")
	parser.add_argument('-p', '--prefix', help="Prefix of output files, if you so choose.", default=None)
	parser.add_argument('-m', '--max_bound', type=int, help="some of the matrices are unbounded stable towards one end, this is the limit the user imposes", default=10)
	parser.add_argument('-n', '--num_sample', type=int, help="number of points to sample when looking for the region of asymptotic stability of the matrix", default=1000)
	parser.add_argument('-z', '--zero_perturb', action='store_true', help="Flag to indicated you want to pertub the zero entries.", default=False)

	# read in the arguments
	args = parser.parse_args()
	input_file = os.path.abspath(args.input_file)
	output_folder = args.output_folder
	if output_folder:
		output_folder = os.path.abspath(output_folder)
	prefix = args.prefix
	max_bound = int(args.max_bound)
	num_sample = int(args.num_sample)
	pert_zero = args.zero_perturb
	if not os.access(output_folder, os.W_OK):
		raise Exception("The provided directory %s is not writable." % output_folder)

	if prefix:
		asymp_stab_file = os.path.join(output_folder, prefix + "_asymptotic_stability.npy")
		num_switch_file = os.path.join(output_folder, prefix + "_num_switch_funcs.npy")
	else:
		asymp_stab_file = os.path.join(output_folder, "asymptotic_stability.npy")
		num_switch_file = os.path.join(output_folder, "num_switch_funcs.npy")

	# check for sanity of input parameters
	if not max_bound > 0:
		raise Exception("max_bound must be larger than 0; provided value: %d." % max_bound)
	if not num_sample > 10:
		raise Exception("num_sample must be larger than 10; provided value: %d." % num_sample)

	# read in the input matrix
	A = np.loadtxt(input_file, delimiter=",")
	Ainv = np.linalg.inv(A)
	m, n = A.shape

	# make sure the original matrix is itself asymptotically stable
	if not SS.is_stable(A):
		raise Exception("Sorry, the input matrix is not stable itself (all eigenvalues must have negative real part). Please try again.")

	# compute the intervals of stability
	intervals = np.zeros((m, n, 2))
	# TODO: parallelize this
	for k in range(m):
		for l in range(n):
			if A[k, l] != 0:
				intervals[k, l, :] = NumSwitch.interval_of_stability(A, Ainv, k, l, max_bound=max_bound, num_sample=num_sample)
			elif pert_zero:
				intervals[k, l, :] = NumSwitch.interval_of_stability(A, Ainv, k, l, max_bound=max_bound, num_sample=num_sample)

	# save these
	print("Saving asymptotic stability to: %s" % asymp_stab_file)
	np.save(asymp_stab_file, intervals)

	# Compute the num switch functions
	# TODO: parallelize this
	crit_epsilon_array = np.zeros((n, n, n, n))
	for k in range(n):
		for l in range(n):
			for i in range(n):
				for j in range(n):
					crit_epsilon_array[k, l, i, j] = NumSwitch.critical_epsilon(Ainv, k, l, i, j)

	num_switch_funcs = dict()
	for k in range(n):
		for l in range(n):
			res = NumSwitch.num_switch_from_crit_eps(crit_epsilon_array, intervals, k, l)
			num_switch_funcs[k,l] = res

	# Save it
	print("Saving shape of num switch functions to: %s" % num_switch_file)
	fid = open(num_switch_file, 'w')
	for k in range(n):
		for l in range(n):
			key = (k,l)
			dict_val = num_switch_funcs[key]
			# TODO: switch based on zero entries
	#for key, dict_val in num_switch_funcs.items():
		#print(key)
		#print(dict_val)
			fid.write("%d\t%d\t" % (key[0], key[1]))
			for i in range(len(dict_val) - 1):
				val, (start, stop) = dict_val[i]
				fid.write("%f\t%f\t%f\t" % (val, start, stop))
			if dict_val:
				val, (start, stop) = dict_val[-1]
				fid.write("%d\t%f\t%f\n" % (val, start, stop))
			else:
				fid.write("\n")
	fid.close()

File: R_version/R/PreprocessMatrix.R
#' Preprocess Matrix
#'
#' This script pre-processes a matrix by figuring out what the intervals of asymptotic
#' stability are, as well as finding which perturbation values lead to a sign switch.
#' @param input_file Input comma separated file for the jacobian matrix.
#' @param output_folder Optional output folder to save python objects to disk. 
#' A number of files will be created in the form ‘output_folder/<prefix>_*.npy’.
#' Default is NULL.
#' @param prefix Prefix of output files, if you so choose.
#' @param max_bound some of the matrices are unbounded stable towards one end, this 
#' is the limit the user imposes. Default: 10
#' @param zero_perturb Flag to indicate you want to perturb the zero entries. 
#' Default: FALSE
#' @param threads Number of threads to use. Default: 1
#' @param verbose Default: FALSE
#' @return A list of with the following objects: matrix_size, column_names, row_names,
#' non_zero, num_switch_functions, asymptotic_stability_start,
#' asymptotic_stability_end, num_switch_funcs_r
#' @export
#' @examples
#' \dontrun{
#' infile <- system.file("extdata", "Modules", "IGP.csv", 
#'     package = "PressPurtCoreAlg") 
#' PreProsMatrix <- PreprocessMatrix(input_file = infile, 
#'     output_folder = NULL, max_bound = 10, threads = 2)
#' }


PreprocessMatrix <- function(input_file, output_folder=NULL, prefix=NULL, max_bound=10, 
                             zero_perturb=FALSE, threads=1, verbose=FALSE){
  # import NumSwitch module and source PreprocessMatrix.py
  reticulate::import_from_path("NumSwitch", 
                               system.file("python", package = "PressPurtCoreAlg"), 
                               convert = F)
  reticulate::source_python(system.file("python", "PreprocessMatrix.py", 
                                        package = "PressPurtCoreAlg"), 
                            convert = F)
  # If output folder is specified, don't output R object
  if(!is.null(output_folder)){
    tt <- py_to_r(run_preproc(input_file, output_folder, prefix, max_bound, 
                              zero_perturb, threads, verbose))
    return(paste0("Output saved to output_folder: ", output_folder)) 
  } else{
    py_temp <- run_preproc(input_file, output_folder, prefix, max_bound, 
                              zero_perturb, threads, verbose)
    tt <- list("original_matrix" = py_to_r(py_temp[[0]]),
               "matrix_size" = py_to_r(py_temp[[1]]),
               "column_names" = py_to_r(py_temp[[2]]),
               "row_names" = py_to_r(py_temp[[3]]),
               "non_zero" = py_to_r(py_temp[[4]]),
               "num_switch_functions_py" = py_temp[[5]],
               "num_switch_functions" = py_to_r(py_temp[[5]]),
               "asymptotic_stability" = py_to_r(py_temp[[6]]))
    #names(tt) <- c("original_matrix", "matrix_size", "column_names", "row_names", "non_zero", 
    #               "num_switch_functions", "asymptotic_stability")
    # separate AS matrix into start and end
    tt$asymptotic_stability_start <- tt$asymptotic_stability[,,1]
    tt$asymptotic_stability_end <- tt$asymptotic_stability[,,2]
    tt$asymptotic_stability <- NULL
    # r based names
    names(tt$num_switch_functions) <- .r_index(names = tt$num_switch_functions, to_r = T)
    # unlist num_switch_funcs
    tt$num_switch_funcs_r <- lapply(names(tt$num_switch_functions), function(x) 
      .NS_func_r(num_switch_funcs = tt$num_switch_functions, name = x))
    names(tt$num_switch_funcs_r) <- names(tt$num_switch_functions)
    tt$num_switch_functions <- NULL
    return(tt)
  }
}
